File: Exercises/101-109 - Small Problems/easy_8/all_substrings.rb
# Write a method that returns a list of all substrings of a string. The returned list should be
#  ordered by where in the string the substring begins. This means that all substrings that start
# at position 0 should come first, then all substrings that start at position 1, and so on.
# Since multiple substrings will occur at each position, the substrings at a given position should
# be returned in order from shortest to longest.

# You may (and should) use the substrings_at_start method you wrote in the previous exercise:

# Examples:

# ```
# substrings('abcde') == [
#   'a', 'ab', 'abc', 'abcd', 'abcde', 
#   'b', 'bc', 'bcd', 'bcde',
#   'c', 'cd', 'cde',
#   'd', 'de',
#   'e'
# ]
# ```

def substrings_at_start(str)
  sub_arr = []
  i = 1

  loop do
    sub_arr << str[0, i]
    i += 1
    break if i > str.size
  end

  sub_arr
end

def substrings(str)
  substrings = []
  i = 0

  loop do
    substrings << substrings_at_start(str[i, str.size])
    i += 1
    break if i >= str.size
  end

  substrings.flatten
end

puts substrings('abcde') == [
      'a', 'ab', 'abc', 'abcd', 'abcde', 
      'b', 'bc', 'bcd', 'bcde',
      'c', 'cd', 'cde',
      'd', 'de',
      'e'
    ]

File: Exercises/210 - Small Problems/list-processing/all-substrings.js
/*
Write a function that returns a list of all substrings of a string. Order the returned list by where in the string the substring begins. This means that all substrings that start at position 0 should come first, then all substrings that start at position 1, and so on. Since multiple substrings will occur at each position, return the substrings at a given position from shortest to longest.

You may (and should) use the substringsAtStart function you wrote in the previous exercise:

Examples:

substrings('abcde');
// result
[ 'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e' ]
  */

function substringsAtStart(str) {
  var subs = [];
  var i;

  for (i = 0; i < str.length; i += 1) {
    subs.push(str.substr(0, i + 1));
  }

  return subs;
}

function substrings(str) {
  var allSubs = [];
  var i;

  for (i = 0; i < str.length; i += 1) {
    allSubs.push(substringsAtStart(str.substring(i)));
  }

  return allSubs.reduce(function(result, arr) {
    return result.concat(arr);
  });
}

console.log(substrings('abcde'));